README.md
# Time left ⏲

[![GoDoc](https://godoc.org/github.com/ismtabo/time-left?status.svg)](https://godoc.org/github.com/ismtabo/time-left) [![Go Report Card](https://goreportcard.com/badge/github.com/ismtabo/time-left)](https://goreportcard.com/report/github.com/ismtabo/time-left)

This is a simple countdown timer that can be used to track the time left until the end of the work day.

![Screenshot](./docs/assets/screenshot.png)

## Usage

To use this timer, simply run the script and it will display the time left until the end of the work day.

```bash
time-left [options]
```

A systray icon will appear in your system tray. Click on it to see the time left until the end of the work day.

![Screenshot](./docs/assets/screenshot.png)

This script can be used to track the time left until the end of the work day. It shows the time left until the end of the work day in the system tray, and the end of the work day. It can be configured to start at a specific time and to last for a specific duration. It can also be configured to take a rest time. See [Configuration](#configuration) for more information.

In addition, by clicking the "Rest" button, the script will add the rest time to the end of the work day. Use the "Reload" button to reload the configuration file and the "Quit" button to close the program.

### Options

The following options are available:

```bash
  -c, --config string   The path to the configuration file (default "$HOME/.config/time-left/config.yaml")
```

## Installation

**Caveat:** Remember to create configuration file before running the script. See [Configuration](#configuration) for more information.

### Requirements

Readmore about systray requirements [here](https://github.com/getlantern/systray#platform-notes).

### Using go get

To install this script, simply run the following command:

```bash
go install github.com/ismtabo/time-left@latest
```

### Source
To install this script, simply clone the repository and build the binary:

```bash
git clone
cd time-left
make build
```

## Configuration

The script can be configured by editing the `config.yaml` file in your XDG config directory (`$HOME/.config/time-left/config.yaml`). 

### Configuration Properties

| Property | Description | Required | Example | Default value |
|:---|:---|:---|:---|:---|
| `start` | The time when the work day starts (HH:mm) | Yes | `09:00` | |
| `duration` | The duration of the work day as a duration string[^1] | Yes | `7h30m` | |
| `rest` | The duration of the rest time as a duration string[^1] | No | `30m` | `1h` |
| `refresh` | The refresh interval as a duration string[^1] | No | `1s` | `30s` |
| `truncate` | The truncate value as a duration string[^1]. This will truncate the time left to the nearest value. See more [time.Truncate](https://pkg.go.dev/time#example-Duration.Truncate).  | No | `30m` | `1m` |

[^1]: **Duration strings**: see more [time.ParseDuration](https://pkg.go.dev/time#ParseDuration).

Use the following snippet to configure the script:

```yaml
# The time when the work day starts
start: 09:00
# The duration of the work day as a duration string
duration: 8h
# (Optional) The duration of the rest time as a duration string
rest: 30m
# (Optional) The refresh interval as a duration string
refresh: 30s
# (Optional) The truncate value as a duration string. This will truncate the time left to the nearest value
truncate: 1m
```

See [config.yaml](./docs/samples/config.yaml) as an example.

## Contributing

To contribute to this project, please fork the repository and submit a pull request.

## License

This project is licensed under the MIT License - see the [LICENSE.md](LICENSE.md) file for details

## Acknowledgements

This project was inspired by the need to track the time left until the end of the work day.

- getlantern/systray: https://github.com/getlantern/systray

## Contributors

- [@ismtabo](https://github.com/ismtabo) (author)


CHANGELOG.md
# Changelog

All notables changes to this project are documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

### Changed

### Deprecated

### Removed

### Fixed

### Security

## [1.1.1] - (2023-08-07)

### Fixed

*   fix: mend `-version` flag, now it shows the version

## [1.1.0] - (2023-07-24)

### Added

*   feat: added end of time day to tray icon
*   feat: added functionality to reload configuration and timer
*   feat: added icons to menu items

## [1.0.2] - (2023-07-20)

### Fixed

*   fix: program was not exiting

## [1.0.1] - (2023-07-20)

### Added

*   feat: add configuration to truncate the time left to the end of the work day
*   feat: add configuration to set refesh interval

### Fixed

*   fix: fix tray icon removes until time when toggling rest flag

## [1.0.0] - (2023-07-20)

### Added

*   feat: add command to see tray icon with time left the end of the work day


icon/icon.go
// File generated by 2goarray v0.1.0 (http://github.com/cratonica/2goarray)

package icon

var Data []byte = []byte{
	0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a, 0x00, 0x00, 0x00, 0x0d,
	0x49, 0x48, 0x44, 0x52, 0x00, 0x00, 0x00, 0x19, 0x00, 0x00, 0x00, 0x19,
	0x08, 0x06, 0x00, 0x00, 0x00, 0xc4, 0xe9, 0x85, 0x63, 0x00, 0x00, 0x00,
	0x09, 0x70, 0x48, 0x59, 0x73, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00,
	0x01, 0x00, 0x4f, 0x25, 0xc4, 0xd6, 0x00, 0x00, 0x03, 0xf2, 0x49, 0x44,
	0x41, 0x54, 0x78, 0x9c, 0xb5, 0x56, 0x7b, 0x48, 0x93, 0x51, 0x14, 0x1f,
	0xa4, 0x52, 0x61, 0x2e, 0x2a, 0x7b, 0x47, 0x05, 0x52, 0x96, 0xd6, 0xb4,
	0x92, 0xec, 0xa1, 0x7f, 0x14, 0xbd, 0xdc, 0xa6, 0x39, 0xd4, 0x99, 0xaf,
	0x05, 0x52, 0x59, 0x44, 0x44, 0xea, 0xfa, 0xa7, 0x22, 0xa5, 0xb2, 0x07,
	0x95, 0x99, 0x96, 0x3d, 0xe6, 0x4a, 0xa3, 0xb2, 0xb7, 0xba, 0xac, 0x48,
	0xcd, 0x1e, 0x14, 0x24, 0xcd, 0xe8, 0x21, 0x44, 0x91, 0x4d, 0x33, 0x4b,
	0x37, 0x4b, 0x2d, 0xcb, 0x72, 0xfb, 0x75, 0xcf, 0x37, 0xa7, 0x9b, 0x7e,
	0x3e, 0x16, 0x74, 0xe1, 0x70, 0xf8, 0xee, 0xfd, 0xbe, 0xf3, 0xfb, 0x7e,
	0xf7, 0xfe, 0xce, 0x39, 0x57, 0x20, 0xe8, 0x7d, 0x08, 0x99, 0xc9, 0x99,
	0xa9, 0x47, 0x0b, 0x87, 0x3e, 0x9b, 0x31, 0x7e, 0x62, 0x9d, 0xcf, 0x64,
	0x37, 0x23, 0xf9, 0x51, 0xce, 0x2e, 0xcf, 0x68, 0xbe, 0x7d, 0x5d, 0xd8,
	0x47, 0x1c, 0xde, 0xe1, 0xc4, 0x2c, 0x71, 0xe2, 0x70, 0x57, 0x5d, 0xe2,
	0xf2, 0x95, 0x28, 0x49, 0x4c, 0x42, 0xf5, 0x81, 0x93, 0x68, 0x48, 0x3b,
	0x8b, 0xaf, 0x69, 0xd9, 0x9c, 0xaf, 0xda, 0x9b, 0x89, 0xdb, 0x71, 0x4a,
	0x6c, 0x9c, 0xe3, 0x8f, 0x71, 0x83, 0x87, 0xe8, 0xe8, 0xfd, 0xf6, 0xef,
	0xfa, 0x35, 0x3c, 0x06, 0x39, 0x3a, 0x3d, 0xdd, 0x21, 0x0d, 0x43, 0xed,
	0x21, 0x15, 0x0c, 0xa9, 0x6a, 0xdc, 0x88, 0x8b, 0x87, 0x72, 0x89, 0x04,
	0x91, 0x73, 0xfd, 0x20, 0xf7, 0xf5, 0xe7, 0xbc, 0x72, 0x89, 0x14, 0xd7,
	0xe3, 0x12, 0xf0, 0x65, 0xff, 0x49, 0xbc, 0xd9, 0xb2, 0x0b, 0x9b, 0xbd,
	0x17, 0x62, 0xe0, 0x00, 0x87, 0x32, 0xfa, 0xbe, 0x2f, 0x00, 0xbf, 0x49,
	0x23, 0x46, 0xea, 0x4b, 0x95, 0xc9, 0xf8, 0x71, 0xec, 0x3c, 0x8e, 0x84,
	0x46, 0x63, 0xbe, 0xbb, 0x08, 0x81, 0x41, 0xd1, 0xd8, 0x77, 0x50, 0x85,
	0xbc, 0x82, 0x07, 0xb8, 0x7a, 0xa3, 0x18, 0xf9, 0x9a, 0x07, 0x38, 0x7c,
	0x34, 0x0b, 0x72, 0x79, 0x0c, 0xfc, 0x3c, 0xbd, 0x90, 0x1a, 0xa6, 0x40,
	0xdd, 0xce, 0x34, 0x5c, 0x91, 0xc6, 0x60, 0x82, 0xb3, 0x50, 0x4f, 0x71,
	0x7a, 0x64, 0xc0, 0x00, 0x0c, 0x2f, 0x93, 0x0f, 0x43, 0xb7, 0xef, 0x38,
	0x02, 0x3c, 0xbd, 0x21, 0x0d, 0x8e, 0x45, 0xd1, 0xbd, 0xf7, 0x68, 0x6a,
	0x04, 0x7e, 0xff, 0x02, 0xea, 0x3e, 0xff, 0x86, 0x2c, 0x38, 0x9c, 0xf3,
	0xf4, 0xfc, 0xb3, 0x05, 0xd0, 0x96, 0x57, 0x22, 0x22, 0x3a, 0x16, 0x62,
	0xd1, 0x6c, 0x54, 0xc4, 0xef, 0x42, 0x71, 0xc8, 0x3a, 0x4c, 0x18, 0x22,
	0x34, 0xf0, 0x31, 0x72, 0x62, 0x5b, 0xa4, 0x25, 0x06, 0x04, 0xe0, 0xeb,
	0xe6, 0x8e, 0xa4, 0x3d, 0x39, 0xd0, 0xd7, 0x03, 0x4d, 0xdf, 0x80, 0x06,
	0xbd, 0x91, 0x99, 0x09, 0x5f, 0x6a, 0x5b, 0x71, 0x29, 0xb7, 0x90, 0xf3,
	0xf4, 0x6c, 0xa8, 0x37, 0x72, 0xeb, 0xa6, 0x36, 0x20, 0x23, 0xf3, 0x1c,
	0xe6, 0x4f, 0x9d, 0x8e, 0x17, 0x1b, 0xb6, 0xe1, 0xb2, 0x38, 0x0a, 0x03,
	0x1d, 0x1c, 0xb4, 0x2c, 0xae, 0xa3, 0x35, 0x88, 0x92, 0xce, 0x80, 0xb6,
	0x48, 0xcc, 0x18, 0x24, 0xa5, 0xe4, 0x20, 0x25, 0x25, 0x1b, 0x9a, 0x82,
	0x52, 0x34, 0x33, 0x16, 0x14, 0xcc, 0x6c, 0x26, 0x7c, 0x6b, 0x00, 0xe7,
	0x2d, 0x73, 0x04, 0xa2, 0xc9, 0x2f, 0x45, 0xe6, 0xf1, 0x1c, 0x64, 0x9c,
	0x38, 0x87, 0x15, 0xd3, 0x45, 0xa8, 0x8c, 0xdb, 0xce, 0x9d, 0x11, 0xc5,
	0xb5, 0x00, 0x08, 0x99, 0x8a, 0x6a, 0xe8, 0x90, 0xe9, 0x0c, 0x02, 0x65,
	0xb1, 0xa8, 0xae, 0x06, 0xf2, 0xae, 0x97, 0x22, 0x23, 0x3d, 0xab, 0x03,
	0xe4, 0xab, 0x01, 0xf8, 0x58, 0xd5, 0x8c, 0xd5, 0x8a, 0x75, 0xcc, 0x7f,
	0xe7, 0x9e, 0x69, 0x9e, 0xd6, 0xd3, 0x8f, 0xaa, 0x39, 0x20, 0x62, 0x14,
	0xa9, 0x58, 0x8b, 0xdd, 0xfe, 0x01, 0xd0, 0x46, 0x6d, 0xc6, 0x78, 0x67,
	0x61, 0x0d, 0x8b, 0xef, 0x42, 0x20, 0xe1, 0x24, 0x53, 0x52, 0xd1, 0x82,
	0x69, 0x22, 0x68, 0x6e, 0xbd, 0x47, 0x75, 0x15, 0xf0, 0xbd, 0x09, 0xed,
	0x00, 0x6d, 0x1d, 0x20, 0xba, 0xca, 0x46, 0x2c, 0x5a, 0xb4, 0x8c, 0xf3,
	0x16, 0x10, 0x5a, 0xa7, 0xf7, 0x88, 0x11, 0x99, 0xf6, 0xf9, 0x07, 0xb6,
	0x6d, 0x1e, 0x78, 0xa5, 0x48, 0xc0, 0xfa, 0x99, 0xf3, 0x88, 0x4d, 0x38,
	0x81, 0x64, 0x53, 0x1e, 0x90, 0x4c, 0xa5, 0x81, 0xd1, 0x78, 0xfb, 0x0e,
	0xf8, 0xfc, 0xc9, 0xc8, 0x9d, 0x43, 0xe7, 0x36, 0x99, 0x41, 0xaa, 0x3e,
	0x34, 0x62, 0xd9, 0x52, 0x31, 0xe7, 0x3b, 0x41, 0x6c, 0x8d, 0xc4, 0x10,
	0xbe, 0x4a, 0x01, 0xd5, 0x62, 0x19, 0x2e, 0x49, 0xa2, 0x08, 0x24, 0x5b,
	0xc0, 0x32, 0xb9, 0x9c, 0x12, 0x6d, 0xeb, 0x52, 0x09, 0xb6, 0x27, 0xab,
	0xa0, 0xd3, 0xd9, 0xee, 0xb9, 0x35, 0x48, 0x6d, 0x4d, 0x0b, 0xc4, 0x01,
	0x2b, 0x39, 0xcf, 0x07, 0x42, 0x62, 0x20, 0xd5, 0xa5, 0x66, 0x9c, 0xc1,
	0x7a, 0x0f, 0x1f, 0x94, 0x45, 0x6c, 0x82, 0xeb, 0x60, 0xe7, 0x72, 0x01,
	0x2b, 0x11, 0xf5, 0x94, 0xc1, 0x91, 0xbe, 0x7e, 0x38, 0xa5, 0xbe, 0x8f,
	0x8a, 0x8a, 0x56, 0x26, 0xd1, 0xd6, 0x6e, 0x4c, 0x68, 0x4b, 0x0a, 0x6f,
	0x3e, 0xc4, 0x98, 0x31, 0xe3, 0x50, 0xc8, 0xf2, 0xc4, 0x56, 0x10, 0x66,
	0x51, 0x90, 0xea, 0x48, 0xde, 0x05, 0x85, 0x8f, 0x20, 0x73, 0xf3, 0xc4,
	0xeb, 0x98, 0x78, 0xb8, 0x0f, 0x1b, 0x59, 0x2f, 0x60, 0xb5, 0xc8, 0x44,
	0xa5, 0x22, 0x6c, 0xae, 0x3f, 0x32, 0x4f, 0x17, 0xb1, 0x3f, 0x95, 0x23,
	0xf7, 0xa2, 0x06, 0x3f, 0x9a, 0x6d, 0x83, 0x18, 0xff, 0x00, 0xa1, 0x21,
	0x91, 0x44, 0x1f, 0x21, 0x21, 0x11, 0xf8, 0xd5, 0x62, 0xbb, 0x4e, 0xef,
	0xe7, 0x5e, 0xd0, 0x70, 0x79, 0x74, 0x2d, 0xff, 0x1e, 0x82, 0xa6, 0xce,
	0x44, 0x05, 0x03, 0x11, 0xb9, 0x8e, 0x45, 0xbf, 0x98, 0x50, 0xc0, 0x92,
	0xe2, 0x32, 0xcc, 0xf2, 0xf6, 0x61, 0x40, 0x11, 0xb8, 0x7b, 0xe7, 0x89,
	0x7d, 0x4c, 0xfa, 0x3a, 0x13, 0x0a, 0x76, 0xe7, 0xf6, 0x63, 0x78, 0x79,
	0xcd, 0x46, 0x49, 0x51, 0x19, 0xc7, 0x88, 0xe6, 0xf4, 0x75, 0x6d, 0xfd,
	0x3f, 0x13, 0x6b, 0x75, 0x49, 0x78, 0xd4, 0x45, 0x01, 0x4f, 0xb0, 0x44,
	0x23, 0xa0, 0xae, 0x5b, 0xd4, 0x93, 0xba, 0xe4, 0xb6, 0xea, 0x3a, 0xdb,
	0x67, 0x9e, 0xd0, 0x1f, 0x93, 0xef, 0xbe, 0x3d, 0x16, 0x6b, 0x5f, 0xb7,
	0xce, 0x93, 0x29, 0x1e, 0x78, 0xdd, 0x25, 0x4f, 0x5c, 0x6c, 0x32, 0x3e,
	0xd8, 0x3a, 0xe3, 0xd5, 0xbd, 0x04, 0xef, 0x64, 0x6a, 0x9b, 0xf1, 0x6b,
	0x78, 0x33, 0xde, 0xae, 0xda, 0x65, 0x56, 0x9d, 0xfd, 0xb5, 0x8b, 0x86,
	0x63, 0x7f, 0xab, 0x30, 0xc9, 0x94, 0xaf, 0x0a, 0xa7, 0x77, 0xad, 0xc2,
	0x03, 0xba, 0x57, 0x61, 0x4b, 0x3f, 0xd1, 0x5b, 0xfa, 0x89, 0xb8, 0xa7,
	0x7e, 0x22, 0xe3, 0xef, 0x27, 0x12, 0xd1, 0x1c, 0xeb, 0x7e, 0xa2, 0xe7,
	0xeb, 0x27, 0x76, 0x75, 0xc6, 0x3c, 0xea, 0x8c, 0x69, 0xaa, 0x7f, 0xea,
	0x8c, 0x1d, 0x8c, 0xcc, 0x3d, 0x3e, 0xf4, 0x5f, 0x7a, 0xfc, 0xd3, 0xde,
	0x18, 0x74, 0x1d, 0x74, 0xeb, 0x48, 0x30, 0xdf, 0x56, 0x82, 0xfe, 0xcb,
	0x6d, 0xc5, 0x7a, 0xf0, 0xdd, 0xbb, 0x4c, 0xf6, 0xde, 0xbb, 0xfe, 0x02,
	0x3e, 0xec, 0x8a, 0x46, 0x7e, 0x6c, 0xd2, 0x7d, 0x00, 0x00, 0x00, 0x00,
	0x49, 0x45, 0x4e, 0x44, 0xae, 0x42, 0x60, 0x82,
}


config/config.go
package config

import (
	"fmt"
	"os"
	"time"

	"gopkg.in/yaml.v3"
)

type Duration time.Duration

func (d *Duration) UnmarshalYAML(value *yaml.Node) error {
	duration, error := time.ParseDuration(value.Value)
	if error != nil {
		return error
	}
	*d = Duration(duration)
	return nil
}

func (d Duration) String() string {
	return time.Duration(d).String()
}

type Config interface {
	// Reload config
	Reload() error
	// Get work day start time
	GetWorkDayStart() time.Time
	// Get work day end time
	GetWorkDayEnd(rest bool) time.Time
	// Get work day duration as duration
	GetWorkDayDuration() time.Duration
	// Get refresh interval
	GetRefreshInterval() time.Duration
	// Get truncate duration
	GetTruncateDuration() time.Duration
}

type config struct {
	path             string
	WorkDayStart     string        `yaml:"start"`
	WorkDayDuration  time.Duration `yaml:"duration"`
	RestDuration     time.Duration `yaml:"rest"`
	RefreshInterval  time.Duration `yaml:"refresh,omitempty"`
	TruncateInterval time.Duration `yaml:"truncate,omitempty"`
}

func NewConfig(path string) Config {
	config := &config{
		path: path,
	}
	if err := config.Reload(); err != nil {
		panic(err)
	}
	return config
}

func (c *config) Reload() error {
	data, err := os.ReadFile(c.path)
	if err != nil {
		return err
	}
	err = yaml.Unmarshal(data, c)
	if err != nil {
		return err
	}
	return nil
}

func (c *config) GetWorkDayStart() time.Time {
	startHour, err := time.Parse("15:04", c.WorkDayStart)
	if err != nil {
		panic(fmt.Sprintf("Invalid start time: %s", c.WorkDayStart))
	}
	now := time.Now()
	return time.Date(now.Year(), now.Month(), now.Day(), startHour.Hour(), startHour.Minute(), startHour.Second(), startHour.Nanosecond(), now.Location())
}

func (c *config) GetWorkDayEnd(rest bool) time.Time {
	if rest {
		return c.GetWorkDayStart().Add(c.WorkDayDuration + c.RestDuration)
	}
	return c.GetWorkDayStart().Add(c.WorkDayDuration)
}

func (c *config) GetWorkDayDuration() time.Duration {
	return c.WorkDayDuration
}

func (c *config) GetRefreshInterval() time.Duration {
	if c.RefreshInterval == 0 {
		return 30 * time.Second
	}
	return c.RefreshInterval
}

func (c *config) GetTruncateDuration() time.Duration {
	if c.TruncateInterval == 0 {
		return 1 * time.Minute
	}
	return c.TruncateInterval
}

func (c *config) String() string {
	return "Config{" +
		"workDayStart='" + c.WorkDayStart + "'" +
		", workDayDuration='" + c.WorkDayDuration.String() + "'" +
		"}"
}


scripts/changelog.mjs
import caporal from "@caporal/core";
import { dedent } from "ts-dedent";
import dayjs from "dayjs";
import { valid } from "semver";

import createVersionChangelog from "./commands/create-version-changelog.mjs";
import getVersionChangelog from "./commands/get-version-changelog.mjs";

caporal.program
  .command(
    "get",
    dedent`Get changelog from version <version>.

    Get changelog sections from version <version> and outputs them.
  
    Example:
    
    $ update-changelog-version get 1.2.3
    
    From:
    
      # Changelog
      
      ## [Unreleased]
      
      ### Added
      
      - Something
      
      ## [1.2.3] - (2021-01-01)
      
      ### Added
      
      - Other thing
      
    Produces:
      
      ## [1.2.3] - (2021-01-01)
      
      ### Added
      
      - Other thing
    `
  )
  .argument("<version>", "Version to get changelog from", { validator: (value) => valid(value) })
  .action(getVersionChangelog)
  .command(
    "create",
    dedent`Update CHANGELOG.md to version <version> with date <date>
  
    Replaces the [Unreleased] section with a new version section and adds a new [Unreleased] section at the top.

    Adds a new version section at the top if it does not exist yet, with name [<version>] - (<date>). Eg: [1.2.3] (2021-01-01)

    Removes empty changes subsections.

    Example:

    $ update-changelog-version 1.2.3 2021-01-01

    From:

      # Changelog

      ## [Unreleased]

      ### Added

      - Something

      ## [1.2.2] - (2020-12-31)

      ### Added

      - Other thing

    Produces:

      # Changelog

      ## [Unreleased]

      ### ...

      ## [1.2.3] - (2021-01-01)

      ### Added

      - Something

      ## [1.2.2] - (2020-12-31)

      ### Added

      - Other thing
    `
  )
  .argument("<version>", "Version to update", { validator: (value) => valid(value) })
  .argument("<date>", "Date to update", {
    default: dayjs(),
    validator: (value) => dayjs(value),
  })
  .action(createVersionChangelog);

caporal.program.run();


scripts/commands/get-version-changelog.mjs
import { remark } from "remark";
import { read } from "to-vfile";
import { find } from "unist-util-find";
import { findAfter } from "unist-util-find-after";
import { findAllAfter } from "unist-util-find-all-after";
import findAllBetween from "unist-util-find-all-between";
import { convert } from "unist-util-is";
import { reporter } from "vfile-reporter";

import { isVersionSection, parseVersionHeader } from "./utils/unist.mjs";

const isVersionSectionForSection = function (version) {
  return convert((node) => {
    if (!isVersionSection(node)) return false;
    const sectionVersion = parseVersionHeader(node.children[0].value);
    return sectionVersion && sectionVersion.version === version;
  });
};

export default async function main({ logger, args, options }) {
  logger.info(`Getting CHANGELOG.md section for ${args.version}`);
  const version = args.version;
  const changelog = await read(resolve(process.cwd(), options.changelog));
  const versionLog = await remark().use(remarkExtractVersionSection, version).process(changelog);
  if (versionLog.messages.length > 0) {
    logger.error(reporter(versionLog.messages));
  }
  // eslint-disable-next-line no-console
  console.log(versionLog.value.replace("\\[", "["));
}

function remarkExtractVersionSection(version) {
  return function transformer(tree, file) {
    const versionSection = find(tree, isVersionSectionForSection(version));
    if (!versionSection) {
      file.fail(`Could not find section for version ${version}`, {
        source: "remark-extract-version-section",
      });
    }
    const previousVersionSection = findAfter(tree, versionSection, isVersionSection);
    const versionChangeSection = previousVersionSection
      ? findAllBetween(tree, versionSection, previousVersionSection)
      : findAllAfter(tree, versionSection);
    return {
      type: "root",
      children: [versionSection, ...versionChangeSection],
    };
  };
}


main.go
package main

import (
	"flag"
	"fmt"
	"log"
	"os"
	"path"
	"time"

	"github.com/adrg/xdg"
	"github.com/getlantern/systray"
	"github.com/ismtabo/time-left/config"
	"github.com/ismtabo/time-left/icon"
)

const (
	TimeLeftFormat = "⏳ Time Left: %s"
	RestTimeFormat = "🍽 Rest: %s"
	EndTimeFormat  = "%s End: %s"
	ReloadText     = "🔄 Reload"
	QuitText       = "❌ Quit"
)

var (
	conf    config.Config
	rest    bool
	done    chan bool
	quit    chan bool
	running bool
)

func main() {
	defaultConfPath := path.Join(xdg.ConfigHome, "time-left", "config.yaml")
	var confOpt string
	version := flag.Bool("version", false, "Print version information")
	flag.StringVar(&confOpt, "config", defaultConfPath, "Path to the config file")
	flag.Parse()
	conf = config.NewConfig(confOpt)
	if version != nil && *version {
		fmt.Printf("Version: %s\n", config.GetVersion())
		fmt.Printf("Build.Revision: %s\n", config.GetRevision())
		fmt.Printf("Build.Time: %s\n", config.GetBuildTime())
		fmt.Printf("Build.OS: %s\n", config.GetOS())
		return
	}
	log.Printf("Version: %s", config.GetVersion())
	log.Printf("build.Revision: %s", config.GetRevision())
	log.Printf("build.Time: %s", config.GetBuildTime())
	log.Printf("build.OS: %s", config.GetOS())
	log.Printf("Config path: %s", confOpt)
	log.Println("Starting...")
	systray.Run(onReady, onExit)
}

func onExit() {
	// clean up here
	log.Println("Exiting...")
	endTimer()
	close(done)
	log.Println("Waiting for goroutines to finish...")
	quit <- true
	close(quit)
	log.Println("Bye!")
	os.Exit(0)
}

func onReady() {
	done = make(chan bool, 1)
	quit = make(chan bool, 1)

	endOfWorkDay := conf.GetWorkDayEnd(rest)
	systray.SetIcon(icon.Data)
	systray.SetTooltip("Time left until the end of the work day")
	mTimeLeft := systray.AddMenuItem("Time left", "Time left")
	mRest := systray.AddMenuItemCheckbox(fmt.Sprintf(RestTimeFormat, "disabled"), "Rest: click to toggle", false)
	mEnd := systray.AddMenuItem(getEndOfWorkdayTitle(endOfWorkDay), "End of the work day")
	mReload := systray.AddMenuItem(ReloadText, "Reload config")
	systray.AddSeparator()
	mQuit := systray.AddMenuItem(QuitText, "Quit the whole app")

	mQuit.SetIcon(icon.Data)

	go func() {
		for {
			select {
			case <-quit:
				return
			case <-mQuit.ClickedCh:
				systray.Quit()
			case <-mRest.ClickedCh:
				rest = !rest
				if rest {
					mRest.SetTitle(fmt.Sprintf(RestTimeFormat, "enabled"))
				} else {
					mRest.SetTitle(fmt.Sprintf(RestTimeFormat, "disabled"))
				}
				mEnd.SetTitle(getEndOfWorkdayTitle(conf.GetWorkDayEnd(rest)))
				if updateTimeLeftMenuItem(mTimeLeft, time.Now(), conf.GetWorkDayEnd(rest)) {
					done <- true
				}
			case <-mReload.ClickedCh:
				reloadConfig(mEnd, mTimeLeft)
			}
		}
	}()

	reloadConfig(mEnd, mTimeLeft)
}

func reloadConfig(mEnd *systray.MenuItem, mTimeLeft *systray.MenuItem) {
	if err := conf.Reload(); err != nil {
		panic(err)
	}
	endOfWorkday := conf.GetWorkDayEnd(rest)
	mEnd.SetTitle(getEndOfWorkdayTitle(endOfWorkday))
	startTimer(endOfWorkday, mTimeLeft)
	running = true
}

func startTimer(wordayEnd time.Time, mTimeLeft *systray.MenuItem) {
	if running {
		done <- true
	}
	ticker := time.NewTicker(1 * time.Second)
	go func() {
		for {
			select {
			case <-done:
				return
			case <-ticker.C:
				if updateTimeLeftMenuItem(mTimeLeft, time.Now(), conf.GetWorkDayEnd(rest)) {
					endTimer()
				}
			}
		}
	}()
	if updateTimeLeftMenuItem(mTimeLeft, time.Now(), wordayEnd) {
		endTimer()
	}
}

func endTimer() {
	if !running {
		return
	}
	done <- true
	running = false
}

func updateTimeLeftMenuItem(item *systray.MenuItem, t time.Time, end time.Time) bool {
	if t.After(conf.GetWorkDayEnd(rest)) {
		item.SetTitle(fmt.Sprintf(TimeLeftFormat, "none. See you tomorrow!"))
		return true
	}
	timeLeft := time.Until(end)
	item.SetTitle(fmt.Sprintf(TimeLeftFormat, timeLeft.Truncate(conf.GetTruncateDuration()).String()))
	return false
}

func getEndOfWorkdayTitle(end time.Time) string {
	return fmt.Sprintf(EndTimeFormat, getEndOfWorkdayClockEmoji(end), end.Format("15:04"))
}

func getEndOfWorkdayClockEmoji(end time.Time) string {
	nextHalfHour := end.Round(30 * time.Minute).Format("03:04")
	switch nextHalfHour {
	case "00:00":
		return "🕛"
	case "00:30":
		return "🕧"
	case "01:00":
		return "🕐"
	case "01:30":
		return "🕜"
	case "02:00":
		return "🕑"
	case "02:30":
		return "🕝"
	case "03:00":
		return "🕒"
	case "03:30":
		return "🕞"
	case "04:00":
		return "🕓"
	case "04:30":
		return "🕟"
	case "05:00":
		return "🕔"
	case "05:30":
		return "🕠"
	case "06:00":
		return "🕕"
	case "06:30":
		return "🕡"
	case "07:00":
		return "🕖"
	case "07:30":
		return "🕢"
	case "08:00":
		return "🕗"
	case "08:30":
		return "🕣"
	case "09:00":
		return "🕘"
	case "09:30":
		return "🕤"
	case "10:00":
		return "🕙"
	case "10:30":
		return "🕥"
	case "11:00":
		return "🕚"
	case "11:30":
		return "🕦"
	}
	return "🛑"
}
